### ISSUE TITLE
completeCheckoutSession example

### ISSUE THREAD

**Issue by schonhoff** [amzn/amazon-pay-api-sdk-php#6]
Hello,

I'm trying to integrate the api to a self-written shop. But I can't get a value for the chargeId and chargePermissionId in the response after the payer runs the amazonPayRedirectUrl. The constraints array is empty. After checking all my code and after 2-3 times reading the api documentation, I saw the following part in the changelog for version 2.0.0:

> There are also non-subtle workflow changes between v1/ and v2/ as the completeCheckoutSession API call will be required now before funds can be captured. See the API Release notes for more details.

I can't find more details in the API Release notes to see how the payload of the completeCheckoutSession API Call looks like. Can you help me out with more information? I tried to contact the amazon pay seller support but they refuse to help me out and said that I need to manually collect every payment. That is not the case if I use the api calls described in the API manual and in the readme.md file.

Can you provide an example api call for the completeCheckoutSession API call? The Payload is my problem.

**Comment by bjguillot** [comment]
Here is an example snippet using completeCheckoutSession:

```php
    $payload = array(
        'chargeAmount' => array(
            'amount' => $amazonpay_order['amount'],
            'currencyCode' => $amazonpay_order['currency']
        )
    );

    $client = new Amazon\Pay\API\Client($amazonpay_config);
    $result = $client->completeCheckoutSession($_POST['checkoutSessionId'], $payload);
```
Let me know if this is able to help you resolve your issue.

Order of the API calls depends on your particular workflow, but in my testing I did it this way:
amazon.Pay.renderButton to create the checkout session, updateCheckoutSession to set the metadata for the order, and then completeCheckoutSession.

**Comment by schonhoff** [comment]
After testing this I got the following response:

```
[
"status" => 401,
  "method" => "POST",
  "url" => "https://pay-api.amazon.eu/sandbox/v2/checkoutSessions/2550535a-cd05-46b5-918e-2365e1f7ea6b/complete",
  "headers" => [
    0 => "accept:application/json",
    1 => "authorization:CENSORED_SIGNITURE",
    2 => "content-type:application/json",
    3 => "user-agent:amazon-pay-api-sdk-php/2.0.0 (PHP/7.2.10; Windows NT/AMD64/10.0)",
    4 => "x-amz-pay-date:20200429T165855Z",
    5 => "x-amz-pay-host:pay-api.amazon.eu",
    6 => "x-amz-pay-region:eu",
  ],
  "request" => "{"chargeAmount":{"amount":129,"currencyCode":"EUR"}}",
  "response" => "",
  "request_id" => "4100c1da-8957-4951-a27d-5327fc3f754a",
  "retries" => 0,
  "duration" => 289,
]
```

Looks like I'm unauthorized but I did not change anything on my $amazonpay_config. The workflow of my shop is the same as you mentioned. I would send the user back to the amazonPayRedirectUrl after the completeCheckoutSession call. But there is the problem. I can't complete it.

Here is my code for the completeCheckoutSession function:

```
$client = new Client([
   'public_key_id' => config('amazon-pay.public_key_id'), 
   'private_key'   => config('amazon-pay.private_key_path'),
   'sandbox'       => true,
   'region'        => config('amazon-pay.region')
]);
$payload = [
    'chargeAmount' => [
        'amount' => 129,
        'currencyCode' => config('amazon-pay.currency_code')
    ]
];
$result = $client->completeCheckoutSession($checkoutSessionId, $payload);
```

Do you have any idea why my request could be denied from the server?

**Comment by bjguillot** [comment]
Are you using delegated access/passing an x-amz-pay-authtoken token header on your other API calls?  If so, you'll need to pass the x-amz-pay-authtoken header on the completeCheckoutSession API call as well, just do something like:

```
$headers = array('x-amz-pay-authtoken' => $config['auth_token']);
$result = $client->completeCheckoutSession($checkoutSessionId, $payload, $headers);
```
If you aren't using x-amz-pay-authtoken anywhere else in your code, then it will be something else that we'll need to dive a little deeper into.

**Comment by bjguillot** [comment]
Hi, I think you have run into an issue on the Amazon side.  I'm seeing a 401 in my test integration as well.  Looking into it.

**Comment by schonhoff** [comment]
Hey,
the only header I used is this:
```
$headers = ['x-amz-pay-idempotency-key' => uniqid()];
```
I hope you find something that caused the error. Thank you already for your help and time investment.

**Comment by bjguillot** [comment]
Issue should be resolved now.  Give it another try.

**Comment by schonhoff** [comment]
Hey,

I give it another try. I got a 422 error. But that was maybe a workflow problem.

```
[
  "status" => 422
  "method" => "POST"
  "url" => "https://pay-api.amazon.eu/sandbox/v2/checkoutSessions/e633fd08-6e46-48f5-a64b-4ed85c12db40/complete"
  "request" => "{"chargeAmount":{"amount":132.5,"currencyCode":"EUR"}}"
  "response" => "{"reasonCode":"InvalidCheckoutSessionStatus","message":"You tried to call an operation on a Checkout Session that is in a state where that operation is not allowed"}"
]
```

I fixed that issue by changing a bit on my order. Now the order is:

- amazon.Pay.renderButton to create the checkout session
- updateCheckoutSession to set the metadata for the order
- redirect the user to amazonPayRedirectUrl
- completeCheckoutSession

Old order:

- amazon.Pay.renderButton to create the checkout session
- updateCheckoutSession to set the metadata for the order
- completeCheckoutSession
- redirect the user to amazonPayRedirectUrl

After i changed it I got a chargeId finally. Thank you so much for your help! Maybe you wanna set the example of the payload in the readme.md. I think it will help others, too.
